### PR TITLE
ARROW-10541: [C++] Add re2 library to core arrow / ARROW_WITH_RE2

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
@github-actions autotune everything

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -363,6 +363,8 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
 
   define_option(ARROW_WITH_UTF8PROC
                 "Build with support for Unicode properties using the utf8proc library" ON)
+  define_option(ARROW_WITH_RE2
+                "Build with support for regular expressions using the re2 library" ON)
```
Comment (nealrichardson) [inline cpp/cmake_modules/DefineOptions.cmake:367]: I find this confusing. I understand how it works--if ARROW_COMPUTE is on, these are on too unless you set them to OFF--but it looks like they're always built by default, unlike the compression libs listed above.

Maybe it's irrelevant that it looks confusing here. Or maybe at least we should add a comment here explaining how the default is really ON iff ARROW_COMPUTE is ON too. Or is it possible to define the default option here to be literally whatever ARROW_COMPUTE is, like `define_option(ARROW_WITH_RE2 "..." ARROW_COMPUTE)`?

Comment (xhochy) [reply cpp/cmake_modules/DefineOptions.cmake:367]: This is sadly the prettiest solution I can think of at the moment. You can define conditional options in CMake but they only make sense based on attributes that are not set by the user but the environment (e.g. which operating system). 

I'll add a comment to this and the above option explaining that they are only used in certain situations.

Comment (bkietz) [reply cpp/cmake_modules/DefineOptions.cmake:369]: Instead of a comment could that explanation be a parenthesis of the description for this option?

Comment (xhochy) [reply cpp/cmake_modules/DefineOptions.cmake]: @bkietz Done 👍

**Comment by xhochy** [comment]
Everything failing in https://github.com/apache/arrow/pull/8759 is past the CMake additions I did here, so this should be good to go.

**Comment by maartenbreddels** [comment]
#8459 builds on Travis, looking good 👍

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -330,6 +330,10 @@ if(ARROW_BUILD_BENCHMARKS
   set(ARROW_TESTING ON)
 endif()
 
+if(ARROW_GANDIVA)
+  set(ARROW_WITH_RE2 ON)
+endif()
```
Comment (kou) [inline cpp/CMakeLists.txt:335]: How about forcing to enable/disable `ARROW_WITH_RE2` in `ThridpartyToolchain.cmake` like `ARROW_WITH_THRIFT`?

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -746,6 +750,14 @@ if(ARROW_WITH_UTF8PROC)
   endif()
 endif()
 
+if(ARROW_WITH_RE2)
+  list(APPEND ARROW_LINK_LIBS RE2::re2)
+  list(APPEND ARROW_STATIC_LINK_LIBS RE2::re2)
+  if(utf8proc_SOURCE STREQUAL "SYSTEM")
```
Comment (kou) [inline cpp/CMakeLists.txt:756]:
```suggestion
  if(RE2_SOURCE STREQUAL "SYSTEM")
```

File: cpp/cmake_modules/DefineOptions.cmake [hunk 4]
```diff
@@ -362,7 +362,9 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
   define_option(ARROW_WITH_ZSTD "Build with zstd compression" OFF)
 
   define_option(ARROW_WITH_UTF8PROC
-                "Build with support for Unicode properties using the utf8proc library" ON)
+                "Build with support for Unicode properties using the utf8proc library (only used if ARROW_COMPUTE is ON)" ON)
+  define_option(ARROW_WITH_RE2
+    "Build with support for regular expressions using the re2 library (only used if ARROW_COMPUTE or ARROW_GANDIVA is ON)" ON)
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake:367]: We can remove `ARROW_WITH_UTF8PROC` and `ARROW_WITH_RE2` if we set them in `ThirdpartyToolchain.cmake` internally like `ARROW_WITH_THRIFT`.

Comment (xhochy) [reply cpp/cmake_modules/DefineOptions.cmake:367]: No, the difference to `ARROW_WITH_THRIFT ` is that you can still build `ARROW_COMPUTE` even if `ARROW_WITH_UTF8PROC=OFF`. In that case, only some features of `compute` won't be build but the majority will work.

**Comment by xhochy** [comment]
@github-actions autotune

File: cpp/cmake_modules/DefineOptions.cmake (hunk 4 above)
Comment (kou) [reply cpp/cmake_modules/DefineOptions.cmake]: I understand.

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**